packages/runtime/container-runtime/src/gc/gcUnreferencedStateTracker.ts

```diff
@@ -45,54 +47,90 @@ export class UnreferencedStateTracker {
 		private readonly inactiveTimeoutMs: number,
 		/** The current reference timestamp used to track how long this node has been unreferenced for. */
 		currentReferenceTimestampMs: number,
-		/** The time after which node transitions to SweepReady state; undefined if session expiry is disabled. */
-		private readonly sweepTimeoutMs: number | undefined,
+		/** The time after which node transitions to TombstoneReady state; undefined if session expiry is disabled. */
+		private readonly tombstoneTimeoutMs: number | undefined, //* Rename to tombstoneTimeoutMs
+		/** The delay from TombstoneReady to SweepReady (only applies if tombstoneTimeoutMs is defined) */
+		private readonly tombstoneSweepDelayMs: number,
 	) {
-		if (this.sweepTimeoutMs !== undefined) {
-			assert(
-				this.inactiveTimeoutMs <= this.sweepTimeoutMs,
-				0x3b0 /* inactive timeout must not be greater than the sweep timeout */,
-			);
-		}
+		assert(
+			this.tombstoneTimeoutMs === undefined ||
+				this.tombstoneTimeoutMs >= this.inactiveTimeoutMs,
+			0x3b0 /* inactiveTimeoutMs must not be greater than the tombstoneTimeoutMs */,
+		);
 
 		this.sweepTimer = new TimerWithNoDefaultTimeout(() => {
 			this._state = UnreferencedState.SweepReady;
+			assert(
+				!this.inactiveTimer.hasTimer && !this.tombstoneTimer.hasTimer,
+				0x3b1 /* inactiveTimer or tombstoneTimer still running after sweepTimer fired! */,
+			);
+		});
+
+		this.tombstoneTimer = new TimerWithNoDefaultTimeout(() => {
+			this._state = UnreferencedState.TombstoneReady;
 			assert(
 				!this.inactiveTimer.hasTimer,
-				0x3b1 /* inactiveTimer still running after sweepTimer fired! */,
+				0x3b1 /* inactiveTimer still running after tombstoneTimer fired! */,
 			);
+
+			if (this.tombstoneSweepDelayMs > 0) {
+				// After the node becomes tombstone ready, start the sweep timer after which the node will be ready for sweep.
+				this.sweepTimer.restart(this.tombstoneSweepDelayMs);
+			} else {
+				this._state = UnreferencedState.SweepReady;
+			}
 		});
 
 		this.inactiveTimer = new TimerWithNoDefaultTimeout(() => {
 			this._state = UnreferencedState.Inactive;
 
-			// After the node becomes inactive, start the sweep timer after which the node will be ready for sweep.
-			if (this.sweepTimeoutMs !== undefined) {
-				this.sweepTimer.restart(this.sweepTimeoutMs - this.inactiveTimeoutMs);
+			// After the node becomes inactive, start the tombstone timer after which the node will be ready for tombstone.
+			if (this.tombstoneTimeoutMs !== undefined) {
+				this.tombstoneTimer.restart(this.tombstoneTimeoutMs - this.inactiveTimeoutMs);
 			}
 		});
+
 		this.updateTracking(currentReferenceTimestampMs);
 	}
 
 	/* Updates the unreferenced state based on the provided timestamp. */
 	public updateTracking(currentReferenceTimestampMs: number) {
```

Comment:
Side-note: I don't think this should be called every time GC runs (which is what happens today). This should only be called during initialization and maybe during transition to write.
Maybe that will make some of the code here simpler.

Reply:
True, once the timers are set it's overkill.  I'll consider that in my follow-up PR to make some changes to this file I have in mind (e.g. we don't need 3 different timer objects, only 1 is ever running at a time)